internal/controller/supervisorconfig/activedirectoryupstreamwatcher/active_directory_upstream_watcher.go

```diff
@@ -0,0 +1,352 @@
+// Copyright 2021 the Pinniped contributors. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package activedirectoryupstreamwatcher implements a controller which watches ActiveDirectoryIdentityProviders.
+package activedirectoryupstreamwatcher
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/klog/v2/klogr"
+
+	"go.pinniped.dev/generated/latest/apis/supervisor/idp/v1alpha1"
+	pinnipedclientset "go.pinniped.dev/generated/latest/client/supervisor/clientset/versioned"
+	idpinformers "go.pinniped.dev/generated/latest/client/supervisor/informers/externalversions/idp/v1alpha1"
+	pinnipedcontroller "go.pinniped.dev/internal/controller"
+	"go.pinniped.dev/internal/controller/conditionsutil"
+	"go.pinniped.dev/internal/controller/supervisorconfig/upstreamwatchers"
+	"go.pinniped.dev/internal/controllerlib"
+	"go.pinniped.dev/internal/oidc/provider"
+	"go.pinniped.dev/internal/upstreamldap"
+)
+
+const (
+	activeDirectoryControllerName = "active-directory-upstream-observer"
+
+	// Default values for active directory config.
+	defaultActiveDirectoryUsernameAttributeName = "userPrincipalName"
+	defaultActiveDirectoryUIDAttributeName      = "objectGUID"
+
+	// By default this group name attribute is the sAMAccountName with special mapping.
+	// Each group will look like sAMAccountName + "@" + domain.
+	// For example if your group sAMAccountName is "mammals" and your domain is
+	// "activedirectory.example.com", it would be mammals@activedirectory.example.com.
+	// This is because sAMAccountName is only unique within a domain, not a forest.
+	defaultActiveDirectoryGroupNameAttributeName = "sAMAccountName"
```

Comment:
If I understand correctly...

1. if the user specifies "sAMAccountName" as the group attribute name in the ActiveDirectoryIdentityProvider spec, then they will get the literal value of the sAMAccountName attribute (no `@domain` appended)
2. if the specifies "" (empty string) as the group attribute name in the ActiveDirectoryIdentityProvider spec, then they will get the magic value of sAMAccountName with the `@domain` appended

Is that right? Seems cool.

Is there a unit test for this controller demonstrating when the user explicitly specifies "sAMAccountName" as the group attribute name?

Reply:
Yep, that's right. If someone wanted to override the default with just `sAMAccountName` they could, and the `@domain` suffix will only show up when the group attribute name is left blank.

I will add that unit test, it seems like a good idea.

Reply:
Resolved in 05afae60c2d3dd95e805ea4a2899513493a1e31d